File: 141-linked-list-cycle/linked-list-cycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */


var hasCycle = function(head) {
    
    // let curr = head;
    // let set = new Set();

    // while(curr) {
    //     if(set.has(curr)) return true;
        
    //     set.add(curr);
    //     curr = curr.next;
        
    // }
    // return false;

    //Approach 2 - Floyd's cycle

    if(!head) return false;

    let slow = head;
    let fast = head.next;

    while(slow!=fast){

        if(!fast || !fast.next) return false;
        fast = fast.next.next;
        slow = slow.next;
    }

    return true;

    


};

File: 141-linked-list-cycle/linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {

        ListNode curr = head;
        Set<ListNode> mySet = new HashSet<>();

        while(curr!=null) {
            if(mySet.contains(curr)) return true;

            mySet.add(curr);
            curr = curr.next;
        }
        return false;


        
    }
}